src/executor/mod.rs

```diff
@@ -0,0 +1,303 @@
+pub(crate) mod current_context;
+pub(crate) mod custom_executor;
+#[cfg(feature = "tokio")]
+pub(crate) mod spawn_blocking;
+
+use std::sync::OnceLock;
+
+use current_context::CurrentContextExecutor;
+use custom_executor::{CustomExecutor, CustomExecutorSyncClosure};
+
+use crate::{global_sync_strategy_builder, Error, ExecutorStrategy, GlobalStrategy};
+
+pub(crate) trait ExecuteSync {
+    /// Accepts a sync function and processes it to completion.
+    async fn execute_sync<F, R>(&self, f: F) -> Result<R, Error>
+    where
+        F: FnOnce() -> R + Send + 'static,
+        R: Send + 'static;
+}
+
+fn set_sync_strategy(strategy: SyncExecutor) -> Result<(), Error> {
+    COMPUTE_HEAVY_SYNC_EXECUTOR_STRATEGY
+        .set(strategy)
+        .map_err(|_| {
+            Error::AlreadyInitialized(COMPUTE_HEAVY_SYNC_EXECUTOR_STRATEGY.get().unwrap().into())
+        })?;
+
+    log::info!(
+        "initialized compute-heavy future executor strategy - {:#?}",
+        global_sync_strategy()
+    );
+
+    Ok(())
+}
+
+/// Get the currently initialized sync strategy,
+/// or the default strategy for the current feature in case no strategy has been loaded.
+///
+/// See [`SyncExecutorBuilder`] for details on strategies.
+///
+/// # Examples
+///
+/// ```
+/// use vacation::{
+///     global_sync_strategy,
+///     global_sync_strategy_builder,
+///     GlobalStrategy,
+///     ExecutorStrategy
+/// };
+///
+/// # fn run() {
+///
+/// #[cfg(feature = "tokio")]
+/// assert_eq!(global_sync_strategy(), GlobalStrategy::Default(ExecutorStrategy::SpawnBlocking));
+///
+/// #[cfg(not(feature = "tokio"))]
+/// assert_eq!(global_sync_strategy(), GlobalStrategy::Default(ExecutorStrategy::CurrentContext));
+///
+/// global_sync_strategy_builder()
+///         .initialize_current_context()
+///         .unwrap();
+///
+/// assert_eq!(global_sync_strategy(), GlobalStrategy::Initialized(ExecutorStrategy::CurrentContext));
+///
+/// # }
+/// ```
+pub fn global_sync_strategy() -> GlobalStrategy {
+    match COMPUTE_HEAVY_SYNC_EXECUTOR_STRATEGY.get() {
+        Some(strategy) => GlobalStrategy::Initialized(strategy.into()),
+        None => GlobalStrategy::Default(<&SyncExecutor>::default().into()),
+    }
+}
+
+pub(crate) fn get_global_sync_executor() -> &'static SyncExecutor {
+    COMPUTE_HEAVY_SYNC_EXECUTOR_STRATEGY
+        .get()
+        .unwrap_or_else(|| <&SyncExecutor>::default())
+}
+
+/// The stored strategy used to spawn compute-heavy futures.
+static COMPUTE_HEAVY_SYNC_EXECUTOR_STRATEGY: OnceLock<SyncExecutor> = OnceLock::new();
+
+/// The fallback strategy used in case no strategy is explicitly set
+static DEFAULT_COMPUTE_HEAVY_SYNC_EXECUTOR_STRATEGY: OnceLock<SyncExecutor> = OnceLock::new();
+
+#[non_exhaustive]
+pub(crate) enum SyncExecutor {
+    /// A non-op strategy that runs the function in the current context
+    CurrentContext(current_context::CurrentContextExecutor),
```

Comment:
I feel like `Context` is overloaded here—But I guess `CurrentThread` is also overloaded with the current thread runtime?

I think we should consider the name here a little bit—Having not deeply read the code, I actually didn't realize that `CurrentContext` meant it was the "identity executor" that did nothing—I thought it used whatever my current runtime was in some way.

But that isn't _really_ the case, it will literally just execute it directly.

Reply:
> But I guess CurrentThread is also overloaded with the current thread runtime?

Yeah that was my concern

I can just make it `ExecuteDirectly`, that's pretty clear?